[jmestxr] iP

jmestxr opened the pull request:
## `/* ask-prof-duke */`


Too much :shit: to remember? **FRET NOT!** ask-prof-duke is here to manage your woes.


### Get it ready in :three: easy steps:

1. download the JAR file from [here](https://github.com/jmestxr/ip/releases/tag/v0.1).
1. open it.
1. add your tasks.


### Features:

 - [x] Managing tasks
 - [ ]  Managing deadlines (coming soon)
 - [ ] Reminders (coming soon)


#### If you Java programmer, you can use it to practice Java too. Here's the main method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

jmestxr committed:
implement level-1 features

jmestxr committed:
implement Level-2 features

jmestxr committed:
Implement Level-3 features

jmestxr committed:
Implement Level-4 features

jmestxr committed:
Test valid commands

jmestxr committed:
Implement Level-5 features

jmestxr committed:
Implement Level-6 features

jmestxr committed:
Implement enum for Command
Enumerate the different types of commands: add, delete, mark, unmark
etc.

jmestxr committed:
Implement Level-7 features

jmestxr committed:
Implement Level-8 features

jmestxr committed:
Make code more OOP by creating classes

jmestxr committed:
Merge branch 'add-gradle-support'

jmestxr committed:
Add checkstyle plugin to gradle and amend style violations

jmestxr committed:
Add Junit tests for Parser and TaskList

jmestxr committed:
Create JAR file for duke

jmestxr committed:
Add JavaDocs

jmestxr committed:
Implement Level-7 features

jmestxr committed:
Implement Level-8 features

jmestxr committed:
Merge branch 'branch-Level-8'

jmestxr committed:
Amend code to adhere to coding conventions

jmestxr committed:
Implement Level-9 features

jmestxr committed:
Merge branch 'branch-A-JavaDoc'

jmestxr committed:
Merge branch 'branch-A-CodingStandard'

jmestxr committed:
Merge branch 'branch-Level-9'

jmestxr committed:
Implement Level-10 (GUI)

jmestxr committed:
Update duke.jar

jefrai commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,62 @@
+     */
+    public Duke(MainWindow app, Path filePath) {
+        this.app = app;
+        this.ui = new Ui(app);
```
Standard recommends omitting `this.` when it isn't needed - makes the code a little easier to read.

jefrai commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,106 @@
+import java.util.ArrayList;
+import java.util.Scanner;
+
+import task.*;
```
Wildcard imports should be broken up into their classes
(The whole package is indeed being used here, but listing each import still helps provide an easy-to-read reference)

jefrai commented on src/main/java/duke/Launcher.java:
```diff
@@ -0,0 +1,16 @@
+    public static void main(String[] args) {
+        Application.launch(Main.class, args);
+    }
+}
```
Seems this source file ended up with no `\n` at its end - having a line terminator at the end of every line, including the last, is good practice to ensure compilers & other programs parse it correctly.

jefrai commented on src/main/java/command/UnmarkCommand.java:
```diff
@@ -0,0 +1,42 @@
+    /**
+     * Unmarks the task with the task no as done.
+     * @param tasks The existing task list.
+     * @param ui The ui of Duke chat.
```
Extra parameter `ui` in this JavaDoc; and in general there should be an empty line separating the description from the parameters.

jefrai commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,106 @@
+                tasksData.createNewFile();
+            }
+
+            Scanner fileReader = new Scanner(tasksData);
```
Great use of spacing - easy to tell what each block of code does at a glance, which really helps to understand longer methods!

jefrai commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,153 @@
+     * @return The input datetime represented as a LocalDateTime object.
+     * @throws DukeException If the input string is not of the supported format: yyyy-M-d HHmm.
+     */
+    public static LocalDateTime parseDateTime(String dateTimeString) throws DukeException {
```
Great work with the JavaDocs - concise & covers almost all methods!

jefrai reviewed (commented):
Just a bit of nitpicking here.
Overall, looks good to go on the standards side!

jmestxr committed:
Add assertions.
There are no assertions implemented.

Assertions allow for certain assumptions to be verified and allow
potential bugs/errors to be spotted as soon as possible.

Assertions are added:
* to check for null values of variables, for example,
checking if taskList is not null after initialisation from storage.
* to check if task is correctly marked done or undone.

jmestxr committed:
Improve code quality per standards.

Some comments were explaining too much on the 'HOW' rather than just
'WHAT' and 'WHY'.

Rectify these comments.

Add missing JavaDocs on methods and classes.

jmestxr committed:
Merge pull request #1 from jmestxr/branch-A-Assertions

Add assertions.

jmestxr committed:
Merge pull request #2 from jmestxr/branch-A-CodeQuality

Improve code quality per standards.

jmestxr committed:
Set up CI

hj-prk commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,153 @@
+    }
+
+    /**
+     * Undo the pretty-formatted datetime string into the user-input datetime string format.
```
Nice JavaDoc implementation!

hj-prk commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,153 @@
+    public static Command parse(String userCommand) throws DukeException {
+        if (CommandPattern.ADD_TODO.match(userCommand)) {
+            /* add todo task */
+            String taskDesc = userCommand.substring(4).trim();
```
What does taskDesc mean? Might be better to rename or explain what it means

hj-prk commented on src/main/java/task/Deadline.java:
```diff
@@ -0,0 +1,44 @@
+     */
+    public Deadline(String description, LocalDateTime by) {
+        super(description);
+        this.by = by;
```
Better to rename by?

hj-prk reviewed (commented):
It generally looks very nice! Except for some minor naming issues. Very good!!!!

jmestxr committed:
Amend checkstyle violations

jmestxr committed:
Implement support for Undo command.

There is no way to undo a previous command should the user makes a
mistake in calling a command that was unintended.

To provide a better experience, i.e. allow the user to rectify the
mistake, this commit implements support for the Undo command.

A stack is created to keep track of the previous commands executed,
excluding commands that are not applicable (List, Find, Exit, Undo).

The previous command can be obtained by popping the stack.

For relevant commands (Add, Delete, Mark, Unmark), an undo method is
implemented to undo the respective effect from executing those commands.

jmestxr committed:
Amend header comments to per coding standard

jmestxr committed:
Minor changes to GUI and remove commented-out code

jmestxr committed:
Abstract out logic for parseStoredTask from Storage to Parser

jmestxr committed:
Handle case of missing local storage directory

jmestxr committed:
Add Ui.png

jmestxr committed:
Add User Guide

jmestxr committed:
Update README.md

jmestxr committed:
Minor changes to GUI and build gradle

jmestxr committed:
Edit Find command to be case-insensitive

jmestxr committed:
Update Ui.png

jmestxr committed:
Update README.md

jmestxr committed:
Amend code to fix checkstyle violations

jmestxr committed:
Update README.md

jmestxr committed:
Update README.md

jmestxr committed:
Update README.md

jmestxr committed:
Update README.md
